web/app.js

```diff
@@ -1542,6 +1542,33 @@ const PDFViewerApplication = {
     };
     internalEvents.set("updatefromsandbox", updateFromSandbox);
 
+    const visitedPages = new Set();
+    const pageOpen = async ({ pageNumber }) => {
+      let actions = null;
+      if (!visitedPages.has(pageNumber)) {
+        // Get actions only the first time the page is visited
+        const page = await pdfDocument.getPage(pageNumber);
+        actions = await page.getJSActions();
+        visitedPages.add(pageNumber);
+      }
+
+      this._scriptingInstance.scripting.dispatchEventInSandbox({
+        id: "page",
+        name: "PageOpen",
+        pageNumber,
+        actions,
+      });
+    };
+    const pageClose = ({ pageNumber }) => {
+      this._scriptingInstance.scripting.dispatchEventInSandbox({
+        id: "page",
+        name: "PageClose",
+        pageNumber,
+      });
+    };
```

Comment:
To ensure that a "pageclose" event doesn't happen *before* "pageopen", since that doesn't seem reasonable (at least to me), the following should prevent that (based on the suggestion above):
```suggestion
    const pageClose = async ({ pageNumber }) => {
      if (!visitedPages.has(pageNumber)) {
        // Ignore the "pageclose" event if there wasn't a corresponding
        // "pageopen" event first.
        return;
      }
      await visitedPages.get(pageNumber);
      
      if (pdfDocument !== this.pdfDocument) {
        return; // The document was closed while the actions resolved.
      }
      
      await this._scriptingInstance.scripting.dispatchEventInSandbox({
        id: "page",
        name: "PageClose",
        pageNumber,
      });
    };
```

Reply:
`visitedPages` is used to avoid to send actions again and again.
So I adapted the code, anyway thx for pointing this.